### ISSUE TITLE
CNI plugin allocated wrong IP to pod

### ISSUE THREAD

**Issue by gyliu513** [projectcalico/calico#1441]
## Steps to Reproduce (for bugs)

Not sure how to reproduce this, but the problem occurred in my cluster as following:
1. Check `ip route`
```
root@BPILICPMSTR001:/etc/cni/net.d# ip route | grep bird
10.1.22.128/26 via 10.10.25.50 dev tunl0  proto bird onlink
10.1.22.192/26 via 10.10.25.50 dev tunl0  proto bird onlink
10.1.153.64/26 via 10.10.25.49 dev tunl0  proto bird onlink
10.1.153.128/26 via 10.10.25.49 dev tunl0  proto bird onlink
10.1.157.0/26 via 10.10.25.54 dev tunl0  proto bird onlink
blackhole 10.1.204.64/26  proto bird
10.1.215.64/26 via 10.10.25.51 dev tunl0  proto bird onlink
10.1.215.128/26 via 10.10.25.51 dev tunl0  proto bird onlink
```

From above, we can see the ip subnet for `10.10.25.54` is `10.1.157.0/26`, so the pod ip address should be using `157` subnet.

2. Check all pods on the nodes, found some pod using ip `10.1.41.xx`, and those ips cannot be accessible.
```
root@BPILICPMSTR001:/etc/cni/net.d# kubectl get pods -n kube-system  -owide | grep 10.10.25.54
calico-node-amd64-7fmkm                                   2/2       Running            0          36d       10.10.25.54    10.10.25.54
compliance-annotator-800001872-vtkjg                      1/1       Running            0          36d       10.1.157.18    10.10.25.54
config-indexer-1347023967-w58cf                           1/1       Running            0          36d       10.1.157.19    10.10.25.54
config-parser-1694631305-dw4r0                            1/1       Running            1          36d       10.1.157.20    10.10.25.54
elasticsearch-client-2232913921-fptds                     2/2       Running            0          36d       10.1.157.1     10.10.25.54
elasticsearch-data-0                                      1/1       Running            0          36d       10.1.157.4     10.10.25.54
elasticsearch-master-1890707706-hgbb7                     1/1       Running            0          36d       10.1.157.2     10.10.25.54
filebeat-ds-amd64-0b2wp                                   1/1       Running            0          36d       10.1.157.3     10.10.25.54
generic-indexer-3009687665-3rr7j                          1/1       Running            0          36d       10.1.157.22    10.10.25.54
k8s-proxy-10.10.25.54                                     1/1       Running            0          36d       10.10.25.54    10.10.25.54
kafka-0                                                   1/1       Running            2          36d       10.1.157.14    10.10.25.54
live-scan-proxy-1671892984-xv8bv                          1/1       Running            0          36d       10.1.157.21    10.10.25.54
logstash-1288778455-z6s7h                                 1/1       Running            0          36d       10.1.157.5     10.10.25.54
monitoring-exporter-404652915-lm9lq                       1/1       Running            0          36d       10.1.157.6     10.10.25.54
monitoring-grafana-1863394620-358lm                       2/2       Running            0          36d       10.1.157.11    10.10.25.54
monitoring-prometheus-68784602-5kz8n                      2/2       Running            0          36d       10.1.157.10    10.10.25.54
monitoring-prometheus-alertmanager-773566672-rf6fj        2/2       Running            0          36d       10.1.157.9     10.10.25.54
monitoring-prometheus-kubestatemetrics-4043960779-v77wl   1/1       Running            0          36d       10.1.157.8     10.10.25.54
monitoring-prometheus-nodeexporter-amd64-t1srh            1/1       Running            0          36d       10.10.25.54    10.10.25.54
notification-dispatcher-3447010162-t6qdt                  1/1       Running            0          36d       10.1.157.23    10.10.25.54
password-annotator-1109186390-vqnw6                       1/1       Running            1          36d       10.1.157.27    10.10.25.54
reg-crawler-4270014643-tn4g4                              0/1       CrashLoopBackOff   8542       32d       10.10.25.54    10.10.25.54
regpod-default-bmrg-app-ci-3.0.13                         1/1       Running            0          51s       10.1.41.40     10.10.25.54
rootkit-annotator-1527073875-qbmkl                        1/1       Running            0          36d       10.1.157.25    10.10.25.54
sas-api-server-3502826129-z1lqb                           1/1       Running            0          36d       10.1.157.16    10.10.25.54
sas-mgmt-server-3232027545-k3x5p                          1/1       Running            0          36d       10.1.157.17    10.10.25.54
statsd-3834029308-nmzmz                                   1/1       Running            0          36d       10.1.157.15    10.10.25.54
va-annotator-1556189863-h3372                             1/1       Running            0          36d       10.1.157.24    10.10.25.54
va-elasticsearch-0                                        1/1       Running            0          36d       10.1.157.12    10.10.25.54
zookeeper-0                                               1/1       Running            0          36d       10.1.157.13    10.10.25.54
```
3. Why CNI allocating ip which does not belong the host subnet? Can anyone show some comments or insight for this? Thanks!

## Your Environment

* Calico version
* Orchestrator version (e.g. kubernetes, mesos, rkt):
* Operating System and version:
* Link to your project (optional):

**Comment by caseydavenport** [comment]
@gyliu513 how many nodes do you have in your cluster? 

From the routes, it looks like blocks have been advertised from four different nodes, most of which have two blocks advertised except for `10.10.25.54` which only has a single block advertised (10.1.157.0/26).

I'm guessing that pods on that node are getting address like `10.1.41.xx`? I'd also guess that the block is failing to be advertised for some reason (If it was then you'd see a route and traffic flow should work).

Could you check the `calico/node` logs on that node to see if there are any errors, and maybe share them here? You may need to enable BIRD debug logging.

Also, could you confirm that all the strange IPs are being assigned to pods on that node?

**Comment by caseydavenport** [comment]
@gyliu513 somewhat related, how many pods / node are you running in this test? I'm surprised to see two blocks allocated per-node, which indicates there might be a problem cleaning up IP addresses for some reason. 

Could you also share the full contents of the Calico etcd?

Something like this:

```
etcdctl ls --recursive -p /calico | grep -v '/$' | xargs -n1 -I '{}' sh -c "echo; echo {}; etcdctl get {} " | tee etcdctl_get
```

Feel free to DM me on Slack if you'd rather not share here.

**Comment by gyliu513** [comment]
Thanks @caseydavenport for the great support! Please check my answers inline.
```
10.10.25.48 BPILICPMSTR001

10.10.25.49 bpilicpwrk001
10.10.25.50 bpilicpwrk002
10.10.25.51 bpilicpwrk001

10.10.25.54 BPILICPMGMT001
```

So I have five nodes as above, there roles are as following:

- BPILICPMSTR001 is kubernetes master and mainly run some kuernetes components, such as kubelet, controller manager etc and also some other add-on componemets to manage network such as calico etc.
- bpilicpwrk001, bpilicpwrk002 and bpilicpwrk003 are worker nodes, those nodes are used to run some customer workloads. 
- BPILICPMGMT001 is management nodes, management node is mainly used to run some ELL, Monitoring services so as to reduce the workload on master node.

Please NOTE: customer workloads will not be scheduled to the nodes of master and management as both of those two nodes have node taint and my workload will not tolerate those taint.

Both master and management node have only one network block as there are not too many containers running on those machines, but for the three worker nodes, as there are many containers running, so after one block is used up, another block is allocated by calico node to make sure the containers can got IP adress.

The following are ip routers for all nodes

```
root@BPILICPMSTR001:~/cluster# ip route | grep bird
10.1.22.128/26 via 10.10.25.50 dev tunl0  proto bird onlink
10.1.22.192/26 via 10.10.25.50 dev tunl0  proto bird onlink
10.1.153.64/26 via 10.10.25.49 dev tunl0  proto bird onlink
10.1.153.128/26 via 10.10.25.49 dev tunl0  proto bird onlink
10.1.157.0/26 via 10.10.25.54 dev tunl0  proto bird onlink
blackhole 10.1.204.64/26  proto bird
10.1.215.64/26 via 10.10.25.51 dev tunl0  proto bird onlink
```

```
root@bpilicpwrk001:~# ip route | grep bird
10.1.22.128/26 via 10.10.25.50 dev tunl0  proto bird onlink
10.1.22.192/26 via 10.10.25.50 dev tunl0  proto bird onlink
blackhole 10.1.153.64/26  proto bird
blackhole 10.1.153.128/26  proto bird
10.1.157.0/26 via 10.10.25.54 dev tunl0  proto bird onlink
10.1.204.64/26 via 10.10.25.48 dev tunl0  proto bird onlink
10.1.215.64/26 via 10.10.25.51 dev tunl0  proto bird onlink
```

```
root@bpilicpwrk002:~# ip route  | grep bird
blackhole 10.1.22.128/26  proto bird
blackhole 10.1.22.192/26  proto bird
10.1.153.64/26 via 10.10.25.49 dev tunl0  proto bird onlink
10.1.153.128/26 via 10.10.25.49 dev tunl0  proto bird onlink
10.1.157.0/26 via 10.10.25.54 dev tunl0  proto bird onlink
10.1.204.64/26 via 10.10.25.48 dev tunl0  proto bird onlink
10.1.215.64/26 via 10.10.25.51 dev tunl0  proto bird onlink
```

```
root@bpilicpwrk003:~# ip route | grep bird
10.1.22.128/26 via 10.10.25.50 dev tunl0  proto bird onlink
10.1.22.192/26 via 10.10.25.50 dev tunl0  proto bird onlink
10.1.153.64/26 via 10.10.25.49 dev tunl0  proto bird onlink
10.1.153.128/26 via 10.10.25.49 dev tunl0  proto bird onlink
10.1.157.0/26 via 10.10.25.54 dev tunl0  proto bird onlink
10.1.204.64/26 via 10.10.25.48 dev tunl0  proto bird onlink
blackhole 10.1.215.64/26  proto bird
```

```
root@BPILICPMGMT001:~# ip route | grep bird
10.1.22.128/26 via 10.10.25.50 dev tunl0  proto bird onlink
10.1.22.192/26 via 10.10.25.50 dev tunl0  proto bird onlink
10.1.153.64/26 via 10.10.25.49 dev tunl0  proto bird onlink
10.1.153.128/26 via 10.10.25.49 dev tunl0  proto bird onlink
blackhole 10.1.157.0/26  proto bird
10.1.204.64/26 via 10.10.25.48 dev tunl0  proto bird onlink
10.1.215.64/26 via 10.10.25.51 dev tunl0  proto bird onlink
```

BTW: I also found an interesting thing, both of the calico node on master and management node does not work now and they always allocate ip address which does not in the range of block for the node.

You can see that it is the two calico node which does not restart that has the problem.

```
root@BPILICPMSTR001:~/cluster# kubectl get pods -n kube-system -owide | grep calico
calico-node-amd64-4dpbl                                   2/2       Running            0          37d       

10.10.25.48    10.10.25.48
calico-node-amd64-6j67j                                   2/2       Running            26         37d       10.10.25.50    10.10.25.50
calico-node-amd64-7fmkm                                   2/2       Running            0          37d       10.10.25.54    10.10.25.54
calico-node-amd64-l6hc9                                   2/2       Running            23         37d       10.10.25.51    10.10.25.51
calico-node-amd64-v22vh                                   2/2       Running            19         37d       10.10.25.49    10.10.25.49
calico-policy-controller-3149890387-tnl76                 1/1       Running            0          37d       10.10.25.48    10.10.25.48
```

Please check the calico dump as attached.
[calico.etcd.log](https://github.com/projectcalico/calico/files/1524488/calico.etcd.log)

Due to the calico node log is very big and I cannot upload it here, but I did check the log there, and found no error logs, do you have any keywords recommended that I can search in the log?

**Comment by gyliu513** [comment]
> somewhat related, how many pods / node are you running in this test? I'm surprised to see two blocks allocated per-node, which indicates there might be a problem cleaning up IP addresses for some reason.

Why allocating two blocks of subnet will indicate some problems when cleaning up IP address? My cluster also have another issue, the pod to pod is not pingable, but node to pod ip works fine. I can ping a pod ip from host, but cannot ping a pod from another pod.

**Comment by caseydavenport** [comment]
> Why allocating two blocks of subnet will indicate some problems when cleaning up IP address?

It doesn't necessarily indicate a problem, but unless you're running >64 pods per node you'd only get a single block on each node. So, it's normal if you're running lots of pods, but abnormal otherwise. 

It sounds like you're running lots of pods on the worker nodes, so two blocks is expected.

**Comment by caseydavenport** [comment]
@gyliu513 I found something interesting in that etcd dump.

Namely, I can see that the block in question has been assigned to host `BPILICPMGMT001`

```
/calico/ipam/v2/host/BPILICPMGMT001/ipv4/
/calico/ipam/v2/host/BPILICPMGMT001/ipv4/block/
/calico/ipam/v2/host/BPILICPMGMT001/ipv4/block/10.1.41.0-26
```

However, the other block assigned to that host uses a lowercase version of the hostname in etcd:

```
/calico/ipam/v2/host/bpilicpmgmt001/ipv4/
/calico/ipam/v2/host/bpilicpmgmt001/ipv4/block/
/calico/ipam/v2/host/bpilicpmgmt001/ipv4/block/10.1.157.0-26
```

I'm not sure _how_ this happened, but it seems like it might be related to the problem. I suspect that some code is properly treating `bpilicpmgmt001` and `BPILICPMGMT001` equally, but that the comparison being done in the BIRD template is not case-insensitive and so is not advertising the `10.1.41.0/26` route.

This is just a suspicion at this point, I'll need to look more closely at code / logs in order to know for sure.

Specifically, I'd need the full kubelet and calico/node debug logs, and perhaps the CNI network config file.

However I'm going to be at kubecon all week so might not get back to this issue until afterwards.

You may want to try deleting this key from etcd:

```
/calico/ipam/v2/host/BPILICPMGMT001/ipv4/block/10.1.41.0-26
```

And then, delete any pods with an IP in 10.1.41.0/26 and let them be re-created. If it comes back with a lowercase entry, and the route is advertised, then I think my suspicion is likely correct.

**Comment by caseydavenport** [comment]
@gyliu513 could you share the version of Calico you're using?

And perhaps `/opt/cni/bin/calico-ipam -v` as well?

**Comment by gyliu513** [comment]
Thanks @caseydavenport  I was using calico 2.4.1 and here you can get the detailed version info for each component https://docs.projectcalico.org/v2.4/releases/

After deleting keys in etcd, pods with error ip and restart calico node, the cluster come back to normal.

**Comment by caseydavenport** [comment]
Ok, fwiw I think this issue has been solved in Calico v3.0, since all hostnames are automatically converted to lowercase.
